lib/client-assets.php

```diff
@@ -718,3 +720,56 @@ function gutenberg_default_script_modules() {
 // Enqueue stored styles.
 add_action( 'wp_enqueue_scripts', 'gutenberg_enqueue_stored_styles' );
 add_action( 'wp_footer', 'gutenberg_enqueue_stored_styles', 1 );
+
+/**
+ * Access the shared static variable for interactive script modules.
+ *
+ * @param string|null $script_module_id The script module ID to register, or null to get the list.
+ * @return array Associative array of script module ID => true.
+ */
+function gutenberg_interactive_script_modules_registry( $script_module_id = null ) {
+	static $interactive_script_modules = array();
+
+	if ( null !== $script_module_id ) {
+		$interactive_script_modules[ $script_module_id ] = true;
+	}
+
+	return $interactive_script_modules;
+}
+
+/**
+ * Register a script module ID for interactive blocks.
+ *
+ * @param string $script_module_id The script module ID.
+ */
+function gutenberg_register_interactive_script_module_id( $script_module_id ) {
+	gutenberg_interactive_script_modules_registry( $script_module_id );
+}
+
+/**
+ * Get the list of interactive script module IDs.
+ *
+ * @return array Associative array of script module ID => true.
+ */
+function gutenberg_get_interactive_script_module_ids() {
+	return gutenberg_interactive_script_modules_registry();
+}
+
+/**
+ * Adds `data-wp-router-options` attribute to script modules registered as interactive.
+ *
+ * @param array  $args The script module attributes.
+ * @param string $id   The script module ID.
+ * @return array Filtered script module attributes.
+ */
+function gutenberg_script_module_add_router_options_attributes( $args, $id ) {
+	// Check if this script module ID is registered as interactive.
+	$interactive_modules = gutenberg_get_interactive_script_module_ids();
+
+	if ( isset( $interactive_modules[ $id ] ) ) {
+		$args['data-wp-router-options'] = '{ "loadOnClientNavigation": true }';
+	}
+		return $args;
+}
+
+add_filter( 'wp_script_module_attributes', 'gutenberg_script_module_add_router_options_attributes', 10, 2 );
```

Comment:
Is this new filter really needed? Can't the existing one be used?

Reply:
Answered here:

- https://github.com/WordPress/wordpress-develop/pull/10324#discussion_r2440533532